pmd-core/src/main/java/net/sourceforge/pmd/ant/internal/PMDTaskImpl.java

```diff
@@ -106,106 +102,44 @@ private void doTask() {
         RuleSetLoader rulesetLoader = RuleSetLoader.fromPmdConfig(configuration)
                                                    .loadResourcesWith(setupResourceLoader());
 
-        // This is just used to validate and display rules. Each thread will create its own ruleset
-        // Substitute env variables/properties
-        String ruleSetString = project.replaceProperties(rulesetPaths);
-
-        List<String> rulesets = Arrays.asList(ruleSetString.split(","));
-        List<RuleSet> rulesetList = rulesetLoader.loadFromResources(rulesets);
-        if (rulesetList.isEmpty()) {
-            throw new BuildException("No rulesets");
-        }
-        logRulesUsed(rulesetList);
+        List<RuleSet> rules = loadRulesets(rulesetLoader);
 
         if (configuration.getSuppressMarker() != null) {
             project.log("Setting suppress marker to be " + configuration.getSuppressMarker(), Project.MSG_VERBOSE);
         }
 
-        // Start the Formatters
-        for (Formatter formatter : formatters) {
-            project.log("Sending a report to " + formatter, Project.MSG_VERBOSE);
-            formatter.start(project.getBaseDir().toString());
-        }
 
-        // log("Setting Language Version " + languageVersion.getShortName(),
-        // Project.MSG_VERBOSE);
+        @SuppressWarnings("PMD.CloseResource")
+        ViolationCounterListener reportSizeListener = new ViolationCounterListener();
 
-        // TODO Do we really need all this in a loop over each FileSet? Seems
-        // like a lot of redundancy
-        Report errorReport = new Report();
-        int problemCount = 0;
-        final String separator = System.getProperty("file.separator");
+        final List<DataSource> files = new ArrayList<>();
+        final List<String> reportShortNamesPaths = new ArrayList<>();
+        StringJoiner fullInputPath = new StringJoiner(",");
 
         for (FileSet fs : filesets) {
-            List<DataSource> files = new LinkedList<>();
             DirectoryScanner ds = fs.getDirectoryScanner(project);
-            String[] srcFiles = ds.getIncludedFiles();
-            for (String srcFile : srcFiles) {
-                File file = new File(ds.getBasedir() + separator + srcFile);
+            for (String srcFile : ds.getIncludedFiles()) {
+                File file = new File(ds.getBasedir() + File.separator + srcFile);
                 files.add(new FileDataSource(file));
             }
 
             final String commonInputPath = ds.getBasedir().getPath();
-            configuration.setInputPaths(commonInputPath);
-            final List<String> reportShortNamesPaths = new ArrayList<>();
+            fullInputPath.add(commonInputPath);
             if (configuration.isReportShortNames()) {
                 reportShortNamesPaths.add(commonInputPath);
             }
+        }
+        configuration.setInputPaths(fullInputPath.toString());
```

Comment:
Is this something we should backport to master?

Reply:
nobody complained yet :) This would only be a problem, if someone uses multiple filesets. We explicitly say "You can use multiple fileset elements. " (https://pmd.github.io/latest/pmd_userdocs_tools_ant.html).

But I checked the code in PMD6 again - we actually call PMD once per fileset. That's why it's correct to overwrite the inputPaths for each iteration. So, if someone configures multiple filesets, we execute PMD multiple times.

You've now refactored it to first collect all files of all filesets and then do a single PMD execution... I guess that's a fine performance improvement.

> Is this something we should backport to master?

So, no. Not needed.